### PR TITLE
Adding decompressor archiver and .rar support

### PR DESCRIPTION
**Pull request by orz25**

### CODE REVIEW

File: archive_extractor/decompressor.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	cReader, isCompressed, err := compression.NewReader(path)
+	if compression.IsGetReaderError(err) {
+		return archiver_errors.New(err)
+	}
+	if err != nil {
+		return err
+	}
+	if !isCompressed {
+		return errors.New("file is not compressed or the compression method is not supported")
```
Comment (orz25) [inline archive_extractor/decompressor.go:34]: Add file name

File: archive_extractor/decompressor.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	processingFunc func(*ArchiveHeader, map[string]interface{}) error, params map[string]interface{}) error {
+	maxBytesLimit, err := maxBytesLimit(path, dc.MaxCompressRatio)
+	if err != nil {
+		return err
+	}
+	provider := LimitAggregatingReadCloserProvider{
+		Limit: maxBytesLimit,
+	}
+	cReader, isCompressed, err := compression.NewReader(path)
+	if compression.IsGetReaderError(err) {
```
Comment (orz25) [inline archive_extractor/decompressor.go:27]: put this flow under if err!=nil

File: archive_extractor/decompressor.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		return err
+	}
+	if !isCompressed {
+		return errors.New("file is not compressed or the compression method is not supported")
+	}
+	defer cReader.Close()
+	limitingReader := provider.CreateLimitAggregatingReadCloser(cReader)
+	defer limitingReader.Close()
+	f, err := os.Open(path)
```
Comment (orz25) [inline archive_extractor/decompressor.go:39]: handle error

File: archive_extractor/decompressor.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package archive_extractor
+
```
Comment (orz25) [inline archive_extractor/decompressor.go:2]: wrap all the errors

File: archive_extractor/decompressor.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	defer f.Close()
+	fInfo, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	// removing the compression extension since now we have a decompressed file
+	name := strings.TrimSuffix(fInfo.Name(), filepath.Ext(fInfo.Name()))
+	archiveHeader := NewArchiveHeader(limitingReader, name, fInfo.ModTime().Unix(), fInfo.Size())
+	err = processingFunc(archiveHeader, params)
+	if err != nil {
```
Comment (orz25) [inline archive_extractor/decompressor.go:49]: not wrap this error

File: archive_extractor/decompressor_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+func TestDecompressor_ExtractArchive_CompressedFile(t *testing.T) {
+	dc := &Decompressor{}
+	funcParams := params()
+	if err := dc.ExtractArchive("./fixtures/test.txt.xz", processingFunc, funcParams); err != nil {
+		fmt.Print(err.Error())
+		t.Fatal(err)
+	}
+	ad := funcParams["archiveData"].(*ArchiveData)
```
Comment (orz25) [inline archive_extractor/decompressor_test.go]: ad, ok := funcParams["archiveData"].(*ArchiveData) 
require true

File: archive_extractor/rar_archiver.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+func (ra RarArchiver) ExtractArchive(path string,
+	processingFunc func(*ArchiveHeader, map[string]interface{}) error, params map[string]interface{}) error {
+	maxBytesLimit, err := maxBytesLimit(path, ra.MaxCompressRatio)
+	if err != nil {
+		return err
+	}
+	provider := LimitAggregatingReadCloserProvider{
+		Limit: maxBytesLimit,
+	}
+	format := archiver.Rar{}
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver.go:23]: why `format`?

File: archive_extractor/rar_archiver.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		}
+		if err != nil {
+			return err
+		}
+		if !file.FileInfo.IsDir() && !utils.PlaceHolderFolder(file.FileInfo.Name()) {
+			countingReadCloser := provider.CreateLimitAggregatingReadCloser(file.ReadCloser)
+			archiveHeader := NewArchiveHeader(countingReadCloser, file.FileInfo.Name(), file.FileInfo.ModTime().Unix(), file.FileInfo.Size())
+			err = processingFunc(archiveHeader, params)
+			if err != nil {
+				return err
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver.go:47]: Do we need to wrap the errors here too?

File: archive_extractor/rar_archiver_test.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/stretchr/testify/assert"
+	"strings"
+	"testing"
+)
+
+func TestRarArchiver(t *testing.T) {
+	ra := &RarArchiver{}
+	funcParams := params()
+	if err := ra.ExtractArchive("./fixtures/test.rar", processingFunc, funcParams); err != nil {
+		fmt.Print(err.Error())
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver_test.go:14]: use `require.NoError`

File: archive_extractor/rar_archiver_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+		t.Fatal(err)
+	}
+	ad := funcParams["archiveData"].(*ArchiveData)
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver_test.go:17]: use and check `ok` param

File: archive_extractor/rar_archiver_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+	assert.Equal(t, ad.ModTime, int64(1454061977))
+	assert.Equal(t, ad.IsFolder, false)
+	assert.Equal(t, ad.Size, int64(695028))
+}
+
+func TestRarArchiver_NoRarFile(t *testing.T) {
+	// zip file with .rar extension (changed manually)
+	ra := &RarArchiver{}
+	funcParams := params()
+	if err := ra.ExtractArchive("./fixtures/notRarFile.rar", processingFunc, funcParams); err != nil {
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver_test.go:28]: what if 'err == nil'? we need to fail the test....
use `assert.Error`

File: archive_extractor/rar_archiver_test.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+		fmt.Print(err.Error() + "\n")
+		assert.Equal(t, "rardecode: RAR signature not found", strings.Trim(err.Error(), ""))
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver_test.go:30]: why trimming `""`?

File: archive_extractor/rar_archiver_test.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	}
+}
+
+func TestRarArchiverReadAll(t *testing.T) {
```
Comment (liron-shalom) [inline archive_extractor/rar_archiver_test.go:34]: rename to `TestRarArchiver_ExtractArchive`

**Comment by liron-shalom** [comment]
APPROVED.

**Review by liron-shalom — APPROVED** [review]

**Pull request merged**